File: modules/data_provider/locust/providers.py
import base64
from datetime import timezone
import random

from faker import Faker

from modules import (
    constants,
    utils,
    exceptions
)
from modules.resource_data_generator.generators import ResourceMixin
from settings.conf import settings

# This is arbitrary limit. Test cases should not break with change of this limit.
# Parameters which are sensitive to this should define their own min/max limits
LIMIT = 10 ** 6


class Provider:

    def __init__(self, profile=None):
        self.profile = profile

    def get_resource(self, resource):
        return Resource(resource).get_resources(profile=self.profile)

    def get_fake_data(self, config):
        generator_class = FakeData()
        fake_func = generator_class.get_fake_mapper(config)

        ret = None

        if fake_func:
            ret = fake_func(generator_class, config)

        return ret

    def generate_data(self, config):

        data_body = {}

        for item_name, item_config in config.items():
            resource = item_config.get(constants.RESOURCE)

            value = self.get_resource(resource) if resource else self.get_fake_data(item_config)

            if value:
                data_body[item_name] = value

        return data_body

    def format_url(self, url, query_config, path_config):
        path_params = self.generate_data(path_config)
        url = url.format_map(utils.StringDict(**path_params))
        query_params = self.generate_data(query_config)
        return url, query_params


class FakeData:

    def __init__(self):
        self.fake = Faker()

    def get_fake_mapper(self, config):

        item_type = config.get(constants.TYPE)

        if not item_type:
            raise exceptions.ImproperSwaggerException("Item type must be defined")

        item_format = config.get(constants.FORMAT)

        fake_func = self.FAKE_MAP.get((item_type, item_format), None)

        # If it did not match, try to match any format
        if not fake_func:
            fake_func = self.FAKE_MAP.get((item_type, "$any"), None)

        return fake_func

    @staticmethod
    def get_enum(config):
        enum_options = config.get(constants.ENUM, [])

        choice = None

        if enum_options:
            choice = random.choice(enum_options)

        return choice

    @staticmethod
    def get_range(config):
        minimum = config.get(constants.MINIMUM, -LIMIT)
        maximum = config.get(constants.MAXIMUM, LIMIT)

        if config.get(constants.MIN_EXCLUDE, False):
            minimum += 1

        if config.get(constants.MAX_EXCLUDE, False):
            maximum -= 1

        if minimum > maximum:
            raise exceptions.ImproperSwaggerException("Minimum cannot be lower than maximum")

        return minimum, maximum

    def get_integer(self, config):
        return (
            self.get_enum(config)
            or random.randint(*self.get_range(config)) * config.get(constants.MULTIPLE_OF, 1)
        )

    def get_float(self, config):

        # Short-circuit return
        num = self.get_enum(config)
        if num:
            return num

        minimum, maximum = self.get_range(config)
        left_side = random.randint(minimum, maximum)
        right_side = round(random.random(), 2)

        final_number = left_side + right_side

        if final_number > maximum:
            final_number = float(maximum)

        if final_number < minimum:
            final_number = float(minimum)

        return final_number

    @staticmethod
    def get_options(config):
        """
        Swagger supports following options:
            - MinLength
            - MaxLength
            - Pattern

        We are only supporting MaxLength for now, and even then, always generating string of maxLength Char always
        """

        return {
            "maximum": config.get(constants.MAX_LENGTH, 100)    # Arbitrarily set max length
        }

    def get_string(self, config):
        return self.get_enum(config) or self.fake.text(max_nb_chars=self.get_options(config)["maximum"])

    def random_date_time(self, config):
        # Date time between 30 years in past to 30 years in future
        return self.fake.date_time_between(end_date='+30y', tzinfo=timezone.utc)

    def get_date(self, config):
        return self.random_date_time(config).strftime("%Y-%m-%d")

    def get_datetime(self, config):
        return self.random_date_time(config).isoformat()

    def get_password(self, config):
        return self.get_enum(config) or self.fake.password(length=self.get_options(config)["maximum"])

    def get_base64(self, config):
        return base64.b64encode(self.get_string(config))

    def get_binary(self, config):
        return self.get_enum(config) or self.fake.binary(length=self.get_options(config)["maximum"])

    def get_email(self, config):
        return self.get_enum(config) or self.fake.free_email()

    def get_uuid(self, config):
        return self.fake.uuid4()

    def get_boolean(self, config):
        return self.fake.boolean()

    def get_array(self, config):

        item_config = config.get(constants.ITEMS)

        if not item_config:
            raise exceptions.ImproperSwaggerException("Items should be defined for Array type - {}".format(config))

        fake_func = self.get_fake_mapper(item_config)

        min_items = config.get(constants.MIN_ITEMS, 0)
        max_items = config.get(constants.MAX_ITEMS, max(10, min_items+1))

        item_length = random.randint(min_items, max_items)

        fake_items = [fake_func(self, item_config) for _ in range(item_length)]

        is_unique = config.get(constants.UNIQUE_ITEMS, False)

        if is_unique:
            fake_items = set(fake_items)

            # If due to de-duplication, our array count decreases and is less than what is required
            while len(fake_items) < min_items:
                fake_items.add(fake_func(self, item_config))

            fake_items = list(fake_items)

        return fake_items

    def get_object(self, config):

        properties = config.get(constants.PROPERTIES)

        if not properties:
            raise exceptions.ImproperSwaggerException("Properties should be defined for Object - {}".format(config))

        fake_object = {}

        for name, prop_config in properties.items():
            fake_func = self.get_fake_mapper(prop_config)
            if fake_func:
                fake_object[name] = fake_func(self, prop_config)

        return fake_object

    FAKE_MAP = {
        # (Type, format) --> function. None should match to no format. any accepts any format at all
        (constants.INTEGER, None): get_integer,
        (constants.INTEGER, "$any"): get_integer,
        (constants.NUMBER, None): get_float,
        (constants.NUMBER, "$any"): get_float,
        (constants.STRING, None): get_string,
        (constants.STRING, constants.DATE): get_date,
        (constants.STRING, constants.DATE_TIME): get_datetime,
        (constants.STRING, constants.PASSWORD): get_password,
        (constants.STRING, constants.BYTE): get_base64,
        (constants.STRING, constants.EMAIL): get_email,
        (constants.STRING, constants.UUID): get_uuid,
        (constants.BOOLEAN, None): get_boolean,
        (constants.ARRAY, None): get_array,
        (constants.OBJECT, None): get_object
    }


class Resource(ResourceMixin):

    def __init__(self, resource_group_name, *args, items=1, flat_for_single=True, **kwargs):
        """
        :param resource_group_name:
        :param items: Number of resources to fetch
        :param flat_for_single: (Boolean)
            For Single resource: (i.e. items==1)
                True: Simplest format
                False: As Iterator
            For multiple resources:
                It is always false
        """

        # Args/kwargs are there to cover function call with spurious arguments
        # Since Class could be instantiated on run-time with variable number of arguments
        # We don't want to raise error to deal with extra arguments

        self.resource_name = resource_group_name
        self.items = items
        self.flat_result = flat_for_single if items == 1 else False

        self.resources = {}

        self.profiles = self.read_file_from_input(settings.PROFILES_FILE, {})
        self.active_profile = ""

    @property
    def profile_config(self):
        return self.profiles.get(self.active_profile)

    @property
    def profile_resource(self):
        return self.get_profile_resource_name(self.active_profile, self.profile_config)

    def resource_set(self):
        resource_set = self.resources.get(self.resource_name, [])

        if len(resource_set) > self.items:
            resource_set = random.sample(resource_set, self.items)

        return list(resource_set)

    def get_resources(self, profile):

        self.active_profile = profile
        self.resources = self.read_file_from_output(
            self.profile_resource, {}, project_sub_folder=settings.RESOURCES_FOLDER
        )

        resources = self.resource_set()

        if not resources:
            raise exceptions.ResourcesException("Resource Pool Not found for - {}".format(self.resource_name))

        if self.flat_result:
            resources = resources[0]

        return resources

File: modules/data_provider/k6/providers.js
import * as _ from 'js_libs/lodash.js';
import * as faker from 'js_libs/faker.js'
import * as yaml from 'js_libs/yaml.js'

import * as constants from 'js_libs/constants.js'
import * as settings from 'js_libs/settings.js'

/*
        Custom Exception Definitions
 */

class InvalidDataOptionsError extends Error {}

class EmptyResourceError extends Error {}

/*
    End Exceptions Definition
 */

const LIMIT = Math.pow(10, 6);
const MILLISECONDS_IN_YEAR = 86400 * 365.25 * 1000;     // Seconds in day * avg. number of days in years * ms in sec


const FakeData = {

    getFakeMapper: function(config) {

        const itemType = _.get(config, constants.TYPE);

        if (!itemType) {
            new InvalidDataOptionsError("Item type must be defined");
        }

        const itemFormat = _.get(config, constants.FORMAT);

        const map = FakeData.FAKE_MAP();
        let fakeFunc = map[[itemType, itemFormat]];

        // If it did not match, try to match any format
        if (!fakeFunc)
            fakeFunc = map[[itemType, "$any"]];

        return fakeFunc;
    },

    FAKE_MAP: function (){
        // (Type, format) --> function. None should match to no format. any accepts any format at all
        let MAP = {};

        // Defined one by one instead of one go since we want to use composite keys
        MAP[[constants.INTEGER, null]] = FakeData.getInteger;
        MAP[[constants.INTEGER, "$any"]] = FakeData.getInteger;
        MAP[[constants.NUMBER, null]] = FakeData.getFloat;
        MAP[[constants.NUMBER, "$any"]] = FakeData.getFloat;
        MAP[[constants.STRING, null]] = FakeData.getString;
        MAP[[constants.STRING, constants.DATE]] = FakeData.getDate;
        MAP[[constants.STRING, constants.DATE_TIME]] = FakeData.getDateTime;
        MAP[[constants.STRING, constants.PASSWORD]] = FakeData.getPassword;
        MAP[[constants.STRING, constants.BYTE]] = FakeData.getBase64;
        MAP[[constants.STRING, constants.EMAIL]] = FakeData.getEmail;
        MAP[[constants.STRING, constants.UUID]] = FakeData.getUUID;
        MAP[[constants.BOOLEAN, null]] = FakeData.getBoolean;
        MAP[[constants.ARRAY, null]] = FakeData.getArray;
        MAP[[constants.OBJECT, null]] = FakeData.getObject;

        return MAP;
    },

    getInteger: function(config) {
        return FakeData.getEnum(config) ||
            _.random(...FakeData.getRange(config)) * _.get(config, constants.MULTIPLE_OF, 1);
    },

    getFloat: function(config) {
        // Short-circuit return
        const num = FakeData.getEnum(config);
        if (num) {
            return num;
        }

        const minMaxArray = FakeData.getRange(config);
        let minimum = minMaxArray[0], maximum = minMaxArray[1];
        const leftSide = _.random(minimum, maximum);
        const rightSide = +Math.random().toFixed(2);        // Plus to convert it back into integer

        let finalNumber = leftSide + rightSide;

        if (finalNumber > maximum) {
            finalNumber = +maximum.toFixed(2);
        }

        if (finalNumber < minimum) {
            finalNumber = +minimum.toFixed(2);
        }

        return finalNumber
    },

    getString: function(config) {
        return FakeData.getEnum(config) ||
            faker.lorem.text(FakeData.getOptions(config)["maximum"])
    },

    getDate: function(config) {
        const date = FakeData.getRandomDateTime(config);
        return `${date.getUTCFullYear()}-${("0" + (date.getUTCMonth() + 1)).slice(-2)}-${("0" + date.getUTCDate()).slice(-2)}`;
    },

    getDateTime: function(config) {
        return FakeData.getRandomDateTime(config).toISOString();
    },

    getPassword: function(config) {
        return FakeData.getEnum(config) || faker.internet.password(FakeData.getOptions(config)["maximum"]);
    },

    getBase64: function(config) {
        return btoa(FakeData.getString(config));
    },

    getEmail: function(config) {
        return FakeData.getEnum(config) || faker.internet.email();
    },

    getUUID: function(config) {
        return faker.random.uuid();
    },

    getBoolean: function(config) {
        return faker.random.boolean();
    },

    getArray: function(config) {
        const itemConfig = _.get(config, constants.ITEMS);

        if (_.isEmpty(itemConfig)) {
            new InvalidDataOptionsError(`Items should be defined for Array type - ${itemConfig}`)
        }

        const fakeFunc = FakeData.getFakeMapper(itemConfig);

        const minItems = _.get(config, constants.MIN_ITEMS, 0);
        const maxItems = _.get(config, constants.MAX_ITEMS, Math.max(10, minItems + 1));

        let fakeItems = _.range(_.random(minItems, maxItems)).map(_ => fakeFunc(itemConfig));

        if (_.get(config, constants.UNIQUE_ITEMS, false)) {
            fakeItems = new Set(fakeItems);

            // If due to de-duplication, our array count decreases and is less than what is required
            while (fakeItems.length < minItems){
                fakeItems.add(fakeFunc(itemConfig))
            }

            fakeItems = Array.from(fakeItems)
        }

        return fakeItems
    },

    getObject: function(config) {
        const properties = _.get(config, constants.PROPERTIES, {});

        if(_.isEmpty(properties)) {
            new InvalidDataOptionsError(`"Properties should be defined for Object - ${config}`)
        }

        let fakeObject = {};

        _.forEach(properties, function(propConfig, name) {
            const fakeFunc = FakeData.getFakeMapper(propConfig);
            if (fakeFunc) {
                fakeObject[name] = fakeFunc(propConfig);
            }
        });

        return fakeObject
    },

    getRandomDateTime: function(config) {
        // Date time between 30 years in past to 30 years in future (approx.)
        const now = _.now();
        const start = now - MILLISECONDS_IN_YEAR * 30;
        const end = now + MILLISECONDS_IN_YEAR * 30;
        return new Date(_.random(start, end));
    },

    getOptions: function(config) {
        /*
        Swagger supports following options:
            - MinLength
            - MaxLength
            - Pattern

        We are only supporting MaxLength for now, and even then, always generating string of maxLength Char always
        */

        return {
            "maximum": _.get(config, constants.MAX_LENGTH, 100)   // Arbitrarily set max length
        };
    },

    getEnum: function(config) {
        const enumOptions = _.get(config, constants.ENUM, []);
        let choice = null;

        if (!_.isEmpty(enumOptions)) {
            choice = _.sample(enumOptions)
        }

        return choice
    },

    getRange: function(config) {
        let minimum = _.get(config, constants.MINIMUM, -LIMIT);
        let maximum = _.get(config, constants.MAXIMUM, LIMIT);

        if (_.get(config, constants.MIN_EXCLUDE, false)) {
            minimum += 1;
        }

        if (_.get(config, constants.MAX_EXCLUDE, false)) {
            maximum -= 1;
        }

        if (minimum > maximum) {
            new InvalidDataOptionsError("Minimum cannot be lower than maximum");
        }

        return [minimum, maximum];
    }

};


class ResourceProvider {
    constructor(resourceName, items, isFlatForSingle) {
        this.resourceName = resourceName;
        this.items = items || 1;
        this.isFlat = this.items === 1 ? isFlatForSingle : false;

        this.resources = {};

        this.profiles = ResourceProvider.readYAMLFile(settings.PROFILES_FILE);
        this.activeProfile = null;
    }

    get profileResource() {
        return _.get(_.get(this.profiles, this.activeProfile, {}), "resource_file", this.activeProfile + ".yaml")
    }

    static readYAMLFile(relativePath) {
        return yaml.load(relativePath);
    }

    resourceSet() {
        // Several Lodash arguments work only on arrays, so converting here if set
        let resourceSet = [..._.get(this.resources, this.resourceName, [])];

        if (resourceSet.length > self.items) {
            resourceSet = _.sampleSize(resourceSet, self.items);
        }

        return resourceSet;
    }

    getResources(profile) {
        this.activeProfile = profile;

        // Not able to find a short and good module in JS which can work with any OS
        // So this will only work with OS which use / as path separator
        this.resources = ResourceProvider.readYAMLFile(settings.RESOURCES_FOLDER + "/" + this.profileResource);

        let resources = this.resourceSet();

        if (_.isEmpty(resources)) {
            new EmptyResourceError(`Resource Pool not found for ${this.resourceName}`);
        }

        if (this.isFlat) {
            resources = resources[0];
        }

        return resources;
    }
}


export class Provider {

    constructor(profile=null) {
        this.profile = profile;
    }

    static getFakeData(config) {
        const fakeFunc = FakeData.getFakeMapper(config);

        let ret = null;

        if (fakeFunc) {
            ret = fakeFunc(config);
        }

        return ret;
    }

    getResource(resource) {
        const resourceProvider = new ResourceProvider(resource);
        return resourceProvider.getResources(this.profile);
    }

    generateData(config) {

        let dataBody = {};
        const self = this;

        _.forEach(config, function(itemConfig, itemName) {

            const resource = _.get(itemConfig, constants.RESOURCE);

            let value = resource ? self.getResource(resource) : Provider.getFakeData(itemConfig);

            if (value) {
                dataBody[itemName] = value;
            }
        });

        return dataBody;
    }
}
